### ISSUE TITLE
Not able to parse bindings that use the .bind() function

### ISSUE THREAD

**Issue by kohlikohl** [knockout/tko.provider#48]
When using the .bind() function in bindings a SyntaxError is thrown.

An example would be:
```js
<h1 data-bind="text: heading, click: onClick.bind(this, 10)"></h1>
```
this throws: `Expected ')' but got 't'` which essentially means calling `.bind()` without any parameters is fine, but totally defeats the purpose.

You can find an example reproducing the issue [here](http://requirebin.com/?gist=401b29d84e6409e854d1afa2f335184a).

Any ideas what could cause this and how to fix it?
Ideally, the solution would be something other than removing `.bind()` from the binding. 😃 

This also works fine with the standard knockout binding provider.

**Comment by brianmhunt** [comment]
Thanks for reporting @kohlikohl 

The linked bin uses knockout-secure-binding, which is supplanted by tko.provider.

Could you update the example to use tko.provider?

**Comment by kohlikohl** [comment]
Thank you for replying that quickly @brianmhunt .

Could you point me into the right direction of using tko.provider in my project.
It looks like the files in the **dist** folder still relate to knockout-secure-binding. I also can't find tko.provider on npm or bower.

Additionally, I am not sure I want to completely replace knockout with tko at this point.

**Comment by brianmhunt** [comment]
No problem.

To install tko.provider you'd use `npm install knockout/tko.provider` – however, where `knockout-secure-binding` was a plugin for knockout, tko.provider is *the* provider (and provider-manager) for tko.

Which is to say, as tko.provider is not a direct replacement (though I wanted them to share the git history), I probably should've left KSB as an independent project.  I may restore it at some point, but it's likely to become an unmaintained project.

Sorry that's roundabout and probably not directly helpful, but I hope it at least helps explain the situation.

Brian

**Comment by kohlikohl** [comment]
Thanks for the explanation. 👍 

Running `npm install knockout/tko.provider` appears to just pull down the current repo into the node_modules which still includes the legacy knockout-secure-binding.js files.

In the end I cloned this repo and ran `gulp` to build a tko.provider.js file, which I then stripped off the universal loader code and stuffed it into a function that exposes `Provider` on the window and assigning an instance of `Provider()` to knockout:
```
ko.bindingProvider.instance = new Provider();
```
This runs, but sadly doesn't produce any output 😭 

It looks like I am not doing this right, or, as you hinted above, `Provider` is not actually meant to be used as a plugin?
Is tko.provider supposed to be compatible with the current version of knockout?

You can see what I have done in [this jsFiddle](https://jsfiddle.net/kohlikohl/9x1op5bo/1/).

**Comment by brianmhunt** [comment]
Hi @kohlikohl – Thanks for putting the effort in.

The tko.provider is one of several packages that form `tko` (which looks to be the basis for KO 4.0).  You can see how it integrates by looking at https://github.com/knockout/tko/blob/master/index.js

tko.provider is designed to replace Knockout 3.x.  Getting tko.provider to work with Knockout 3.x would probably be possible, but not straightforward because a whole lot of Knockout 3.x had to be duplicated in tko.provider. 😞 

In the future I hope to draw a better delineation between the data-binding (linking DOM nodes to binding handlers) and the parsing of the data-bind string, which could make it possible – or at least easier – for tko.provider to be used in knockout 3.x, but that's a ways off.

**Comment by kohlikohl** [comment]
Ok. So it sounds like **tko.provider** is NOT a replacement for **knockout-secure-binding** at this point. Is that correct?

Circling back to my initial issue with **knockout-secure-bindings**, do you have any idea on how to solve the issue?
Could you take a look at it, or point me in the right direction where to start looking?

**Comment by brianmhunt** [comment]
@kohlikohl Yes, you are correct.  tko.provider – if back-ported for compatibility – would solve the issue.

As to the `onClick.bind`, KSB only supports argument-less functions (being a more "pure" templating language), so likely the best option is to have `onClick` already bound.  That may not always be an option, so here are a couple others:

## For the `this` part:

1. Get the `ko.dataFor(evt.target)`  (or `evt.currentTarget`, as the case may be, and use that as `this`; or
2. Use `ko.punches` and have `on.click: onClick` – the `this` should be assigned correctly.

## For the `10` part:

Add a property to the DOM node (e.g. `data-number="10"`) and get it with `evt.target.getAttribute('data-number')`.

Those are the only things that come to mind immediately.  You might have some luck on Stackoverflow or gitter.im/knockout/knockout, too.

Do either of those help?

**Comment by kohlikohl** [comment]
Thank you for your suggestions @brianmhunt.

I have a few more questions: 

1. You mentioned that if tko.provider would be back-ported to work with the current version of knockout, the issues I am seeing would be resolved. 
Is there any plan to back-port it?

1. Furthermore, since KSB, is a plugin for knockout, should it not be compatible with all forms of bindings that knockout itself supports? e.g. if function calls with arguments work in knockout, shouldn't they also work in KSB?

1. Last but not least, it looks like there is development towards a new version of knockout which will make use of the tko.provider library and therefore inherently be CSP compliant. Is that statement correct?
If so, do you have any timelines as to when the next release of knockout can be expected?

**Comment by brianmhunt** [comment]
Thanks @kohlikohl – To answer your questions:

1. Not right now.

2. KSB essentially replaces bits of the browser's Javascript parser.  Knockout "bound" elements like this: `new Function("with($context)with($data){"+...binding-code..."+}"`.  KSB is not a full-fledged Javascript parser; it only replicates a reasonable subset of what Javascript does, so Knockout could eliminate the `eval`/`new Function`.

3. TKO works well for a many cases where Knockout was once used, but there's more on the TODO (here: github.com/knockout/tko/issues/1) — notably testing in a broader set of environments and documentation.  Depending on availability, we're probably looking at releasing a KO 4.0 beta in Q1 of 2017, with the full release to follow.

I hope that helps. :smile:
